### ISSUE TITLE
The demo code doesn't work properly: Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by naluckstar** [aws/aws-iot-device-sdk-embedded-C#1633]
Both the setups are in the same network. Steps followed -

1.Created a thing in AWS. Download the certificate, private and public key.
2.Activate the keys and attach a policy.
3.Added Endpoint, certificate and private key paths in demo_config.h
4.Compiled and ran it.

The questions are as follows:

root@hlk:/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build# make mqtt_demo_mutual_auth
[ 33%] Built target sockets_posix
[ 66%] Built target openssl_posix
[ 66%] Built target clock_posix
Scanning dependencies of target mqtt_demo_mutual_auth
[100%] Building C object demos/mqtt/mqtt_demo_mutual_auth/CMakeFiles/mqtt_demo_mutual_auth.dir/mqtt_demo_mutual_auth.c.o
[100%] Linking C executable ../../../bin/mqtt_demo_mutual_auth
[100%] Built target mqtt_demo_mutual_auth
root@hlk:/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build# cd bin
root@hlk:/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin# ./mqtt_demo_mutual_auth 
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
Segmentation fault (core dumped)
root@hlk:/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin#

**Comment by yourslab** [comment]
How are you formatting the file paths? Have you tried using an absolute directory?

**Comment by naluckstar** [comment]
1.The file configuration is as follows:

#define ROOT_CA_CERT_PATH          "/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/AmazonRootCA1.crt"
#define CLIENT_CERT_PATH           "/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/89797c3d3e-certificate.pem.crt"
#define CLIENT_PRIVATE_KEY_PATH    "/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/89797c3d3e-private.pem.key"

2.After setting the debug level flag(LOG_DEBUG),then re-build and re-run the compiled program again,the debugging information is as follows:

root@hlk:/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin# ./mqtt_demo_mutual_auth 
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[WARN] [DEMO] [mqtt_demo_mutual_auth.c:649] Connection to the broker failed. Retrying connection after 206 ms backoff.
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[WARN] [DEMO] [mqtt_demo_mutual_auth.c:649] Connection to the broker failed. Retrying connection after 575 ms backoff.
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[WARN] [DEMO] [mqtt_demo_mutual_auth.c:649] Connection to the broker failed. Retrying connection after 768 ms backoff.
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[WARN] [DEMO] [mqtt_demo_mutual_auth.c:649] Connection to the broker failed. Retrying connection after 415 ms backoff.
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[WARN] [DEMO] [mqtt_demo_mutual_auth.c:649] Connection to the broker failed. Retrying connection after 200 ms backoff.
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[ERROR] [DEMO] [mqtt_demo_mutual_auth.c:642] Connection to the broker failed, all attempts exhausted.
[ERROR] [DEMO] [mqtt_demo_mutual_auth.c:1563] Failed to connect to MQTT broker a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com.
[INFO] [DEMO] [mqtt_demo_mutual_auth.c:1580] Short delay before starting the next iteration....

[INFO] [DEMO] [mqtt_demo_mutual_auth.c:628] Establishing a TLS session to a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com:8883.
[DEBUG] [Transport_OpenSSL_Sockets] [openssl_posix.c:179] Attempting to open Root CA certificate: Path=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:316] fopen failed to find the root CA certificate file: ROOT_CA_PATH=/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:627] Setting up credentials failed.
[ERROR] [Transport_OpenSSL_Sockets] [openssl_posix.c:672] Failed to establish a TLS connection.
[WARN] [DEMO] [mqtt_demo_mutual_auth.c:649] Connection to the broker failed. Retrying connection after 71 ms backoff.

**Comment by yourslab** [comment]
I'm a little confused because the log is indicating `/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/certificates/AmazonRootCA1.crt` even though you specified `/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/AmazonRootCA1.crt`. You seem to be missing a `bin` folder?

**Comment by yourslab** [comment]
Also, you shouldn't have to change `ROOT_CA_CERT_PATH`. If you leave it as is, CMake will download it for you.

**Comment by yourslab** [comment]
Hi @naluckstar. Any update on this issue?

**Comment by naluckstar** [comment]
Yes, the problem has been solved. The path needs to be reset. I used this command:
cmake . -Bbuild -DAWS_IOT_ENDPOINT="a241jm1zvbp1tq-ats.iot.us-west-2.amazonaws.com" -DROOT_CA_CERT_PATH="/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/AmazonRootCA1.crt" -DCLIENT_CERT_PATH="/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/89797c3d3e-certificate.pem.crt" -DCLIENT_PRIVATE_KEY_PATH="/home/hlk/Desktop/aws/aws-iot-device-sdk-embedded-C/build/bin/certificates/89797c3d3e-private.pem.key" -DTHING_NAME="arn:aws:iot:us-west-2:154566201365:thing/test"

It seems that this is necessary, especially after the SDK has changed its path.Thank you for your help.

**Comment by yourslab** [comment]
Okay, I'm glad you got it resolved. I will close this now, but please feel free to open another issue if ever you face more issues.
